Audio component

chmorgan opened the pull request:
@tore-espressif here is the updated PR that uses external components and renames the playlist component to I hope better reflect what the component does.

This includes all of the fixes from the previous PR review.

Thoughts on next steps?

tore-espressif reviewed (commented):
@chmorgan , thank you very much for the update.

I'd suggest following next steps: Upload your new components to our IDF component registry. In case you missed it, here are some useful links that will give you more information about how IDF component manager works:
* [Espressif API guide](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/tools/idf-component-manager.html) 
* [PyPi registry](https://pypi.org/project/idf-component-manager/)
* The component service itself is hosted [here](https://components.espressif.com/)

The components can be uploaded either manually or automatically via GitHub actions (such as [here ](https://github.com/lvgl/lvgl/blob/master/.github/workflows/esp_upload_component.yml), I can help you with setting it up.

Then, we defined dependencies on these components in esp-box repository (instead of adding the components as submodules)

tore-espressif commented:
You will need an API token for uploading the components. I can send it to you in private message

chmorgan commented:
@tore-espressif sure, you can email to me via chmorgan@gmail.com, or if you prefer signal or some other encrypted means I can give you my mobile.

chmorgan commented on .gitmodules:
```diff
@@ -10,3 +10,12 @@
+[submodule "components/esp32-file-iterator"]
+	path = components/esp32-file-iterator
+	url = https://github.com/chmorgan/esp32-file-iterator.git
+[submodule "components/esp32-libhelix-mp3"]
+	path = components/esp32-libhelix-mp3
+	url = https://github.com/chmorgan/esp32-libhelix-mp3.git
```
Checked the license, it looks like its compatible as the license is quite permissible. Many of the terms appear intended to prevent patents to be applied ontop of the RealNetworks IP (libhelix-mp3).

chmorgan commented:
@tore-espressif just read your message here, let me get those files in place and get things working/building and ready for the secret.

tore-espressif commented on examples/factory_demo/main/idf_component.yml:
```diff
@@ -0,0 +1,19 @@
+## IDF Component Manager Manifest File
+dependencies:
+  chmorgan/esp-libhelix-mp3: "*"
+  chmorgan/esp-file-iterator: "*"
+  chmorgan/esp-audio-player: "*"
```
Could you please use fixed versions for these components?

tore-espressif commented on examples/mp3_demo/main/idf_component.yml:
```diff
@@ -0,0 +1,19 @@
+## IDF Component Manager Manifest File
+dependencies:
+  chmorgan/esp-libhelix-mp3: "*"
+  chmorgan/esp-file-iterator: "*"
+  chmorgan/esp-audio-player: "*"
```
same here

tore-espressif reviewed (commented):
@chmorgan Thank you for your effort.
One last comment: Could you please use fixed versions of your components? Since we don't maintain these components, we'd like to avoid any breaking changes that could be introduced if you uploaded new versions

Otherwise LGTM!

chmorgan commented:
@tore-espressif switched to fixed versions. It's been a journey but this looks like a good first set of changes to get in.

chmorgan commented on examples/mp3_demo/main/audio.c:
```diff
@@ -1,6 +1,5 @@
 /**
- * @file audio.c
- * @brief 
+ * @file
```
Surprised I forgot that file. Removed it in the commit where it was removed from the CMakeLists.txt file.

tore-espressif commented:
@Real-YuZhe @leeebo @wujiangang PTAL (related code reviews #30 )

espressif2022 commented:
@chmorgan ，hello：
I tested and found the following situations：
when the playback started, or the playback resume from pause, there is no sound from the speaker;
I must click the button "vol +/-" to recovery it.

I notice that，you'll do disable mute after function "aplay_file(i, audio_event.fp)" 
i->config.mute_fn(AUDIO_PLAYER_MUTE);
es8311_write_reg(ES8311_DAC_REG32, 0x00);

the vol register(ES8311_DAC_REG32) will be cleared, need to do some improvement?

espressif2022 commented:
> This should have been addressed in the mute function, the original volume should be restored there when exiting mute. Maybe a change was lost somewhere. Were you testing with mp3 demo or factory demo in the esp-box repository? Chris
> […](#)
> On Mon, Sep 5, 2022 at 7:48 AM espressif2022 ***@***.***> wrote: @chmorgan <https://github.com/chmorgan> ，hello： I tested and found the following situations： when the playback started, or the playback resume from pause, there is no sound from the speaker; I must click the button "vol +/-" to recovery it. I notice that，you'll do disable mute after function "aplay_file(i, audio_event.fp)" i->config.mute_fn(AUDIO_PLAYER_MUTE); es8311_write_reg(ES8311_DAC_REG32, 0x00); the vol register(ES8311_DAC_REG32) will be cleared, need to do some improvement? — Reply to this email directly, view it on GitHub <[#42 (comment)](https://github.com/espressif/esp-box/pull/42#issuecomment-1236896188)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJH4AFAZTMWVBFE3JY6HBDV4XMY7ANCNFSM57U6BCPA> . You are receiving this because you were mentioned.Message ID: ***@***.***>
I'm testing with factory demo；
I've check it, mp3 demo did the operation "es8311_codec_set_voice_volume"
I lost this part of code in factory demo?

chmorgan commented:
@espressif2022 hmm. I fixed a minor nit with the global static, moved it into the mute function, and applied the mp3_demo mute function to the factory_demo (as it looks like you did in testing.) That's what is pushed up right now.

On the "Hi esp" issue, thoughts on how to resolve? The intent of muting is to eliminate noise, clicks, pops etc, that can occur when transitioning to/from playback. I'm still seeing these here however, when playing/pausing and I'm still unsure of the origin of that noise.

We *could* remove the muting functionality from the mute handler entirely. That would keep the codec unmuted (and thus not adjust REG32 to zero like the es8311 set voice mute function does. We could also have other audio playback unmute/play/mute.

Thoughts on the best approach? and I do want to resolve the clicking issue, its on the list of things to look at at some point.

chmorgan commented:
@espressif2022 hmm. Yeah that zero dma was in place before and it didn't address the issue, at the same time I'm not sure the mp3 decode was properly writing data to the buffer.

Mind if I revisit the clicking at a later time? This PR plus components is the culmination of a few months of outstanding work. If you feel this PR and components are moving the project/ecosystem forward I'd like to get that in place and iterate on top of that work with future PRs.

espressif2022 commented:
> @espressif2022 hmm. Yeah that zero dma was in place before and it didn't address the issue, at the same time I'm not sure the mp3 decode was properly writing data to the buffer.
> 
> Mind if I revisit the clicking at a later time? This PR plus components is the culmination of a few months of outstanding work. If you feel this PR and components are moving the project/ecosystem forward I'd like to get that in place and iterate on top of that work with future PRs.

Ok, I will submit to our internal gitlab in the next few days.
This really took a lot of time, I'm very sorry

espressif2022 commented:
@chmorgan ,hello！
We have merged into gitbub last week. You can confirm it when we have time. If it is ok, it will be closed.
Thanks!

chmorgan commented:
Oh ok. I just saw the commits, thank you! Sorry about the stuff that had to
be fixed. I also saw the lvgl 8.3 pr was merged. The plan is to switch that
to a component in the near future as well.



On Sun, Sep 25, 2022 at 11:18 PM espressif2022 ***@***.***>
wrote:

> @chmorgan <https://github.com/chmorgan> ,hello！
> We have merged into gitbub last week. You can confirm it when we have
> time. If it is ok, it will be closed.
> Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/espressif/esp-box/pull/42#issuecomment-1257420717>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJH4AHBFHQ4IPDAFLM3VHDWAEIZ5ANCNFSM57U6BCPA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

espressif2022 commented:
> Oh ok. I just saw the commits, thank you! Sorry about the stuff that had to be fixed. I also saw the lvgl 8.3 pr was merged. The plan is to switch that to a component in the near future as well.
> […](#)
> On Sun, Sep 25, 2022 at 11:18 PM espressif2022 ***@***.***> wrote: @chmorgan <https://github.com/chmorgan> ,hello！ We have merged into gitbub last week. You can confirm it when we have time. If it is ok, it will be closed. Thanks! — Reply to this email directly, view it on GitHub <[#42 (comment)](https://github.com/espressif/esp-box/pull/42#issuecomment-1257420717)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAJH4AHBFHQ4IPDAFLM3VHDWAEIZ5ANCNFSM57U6BCPA> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Thanks for your work.
Lvgl 8.3 adds a lot of features, which are pretty good.
